### ISSUE TITLE
feat: support filtering datasets by sample id

### ISSUE THREAD

**Issue by flymin** [mvp-ai-lab/mvp-dataset#21]
## Problem

Some pretraining datasets need to include or exclude samples by a predefined list of sample ids.

For example, an experiment may use:

- `include_ids_file`: only keep samples whose `id` is listed in the file
- `exclude_ids_file`: drop samples whose `id` is listed in the file

This is useful for train/validation split control, deduplication, and reproducing existing experiment configs.

Currently, `mvp-dataset` has related dataset operations such as:

- `select(fields)`: field projection, not sample filtering
- `sample(fraction)`: random subset by fraction
- `split(fractions)`: row-based split
- Lance `_selection`: row-window / sampled-row selection
- `assemble(...)`: can be used to implement a lazy post-read filter manually

But there is no first-class operation for filtering samples by id.

## Requested Feature

Add a dataset filtering op that supports id-based include/exclude semantics.

Suggested API shape:

```python
dataset = Dataset.from_source(...)
dataset = dataset.filter_by_id(
    id_field="id",
    include_ids_file="/path/to/include.txt",
    exclude_ids_file="/path/to/exclude.txt",
)
```

or a more general filter API:

```python
dataset = Dataset.from_source(...)
dataset = dataset.filter(predicate)
```

with a built-in helper for id files.

## Desired Semantics

- `include_ids_file`: keep only samples whose `id_field` is in the file
- `exclude_ids_file`: drop samples whose `id_field` is in the file
- If both are provided, apply include first, then exclude
- The id file should contain one id per line
- Empty lines should be ignored
- The filter config should be included in dataset fingerprint / resume compatibility checks
- Missing `id_field` should raise a clear error

## Lance-Specific Consideration

For Lance datasets, a source-level implementation would be preferable when possible:

1. Resolve ids to row positions
2. Cache the id-to-row-index mapping or selected row positions
3. Read only selected rows during iteration

This avoids the inefficiency of post-read filtering, especially when the include set is small or the exclude set is large.

A simpler first implementation could be a lazy post-read filter stage, but the long-term goal should be source-level filtering for Lance.

## Motivation

This is needed to reproduce LLM pretraining configs that rely on train/validation split files, such as excluding validation ids from the training dataset and evaluating only on a fixed validation id set.

**Comment by flymin** [comment]
## Request to reopen: sparse include filtering is functionally correct but does not scale                                                                                                                                                                
                                                                                                                                                                                                                                                           
  The implementation merged in #23 provides the expected filtering semantics, caching, ordering, deduplication, and resume compatibility. However, production-scale testing shows that the source-level performance goal described in this issue has not   
  yet been met for sparse ID inclusion over a very large Lance source.                                                                                                                                                                                     
                                                                                                                                                                                                                                                           
  ### Workload                                                                                                                                                                                                                                             
                                                                                                                                                                                                                                                           
  The observed workload has:                                                                                                                                                                                                                               
                                                                                                                                                                                                                                                           
  - approximately 2.95 billion physical rows                                                                                                                                                                                                               
  - 8,323 Lance fragments                                                                                                                                                                                                                                  
  - a valid scalar/BTREE index on the ID column covering all fragments                                                                                                                                                                                     
  - an include file containing 81,000 unique IDs                                                                                                                                                                                                           
  - predicate batches of 512 IDs, following the example in #23                                                                                                                                                                                             
  - 159 bounded `IN (...)` predicates                                                                                                                                                                                                                      
  - 8 cooperating ranks per node                                                                                                                                                                                                                           
  - 64 filter-index build parts                                                                                                                                                                                                                            
                                                                                                                                                                                                                                                           
  The caller follows the intended usage:                                                                                                                                                                                                                   
                                                                                                                                                                                                                                                           
  1. Read and deduplicate IDs.                                                                                                                                                                                                                             
  2. Split them into bounded `IN (...)` predicates.                                                                                                                                                                                                        
  3. Pass the complete predicate list as one OR-combined filter operation.                                                                                                                                                                                 
  4. Use the existing scalar/BTREE index.                                                                                                                                                                                                                  
  5. Allow all ranks to participate in cache preparation.                                                                                                                                                                                                  
                                                                                                                                                                                                                                                           
  ### Observed behavior                                                                                                                                                                                                                                    
                                                                                                                                                                                                                                                           
  Filter-index preparation partitions the source into 64 fragment ranges. For every source partition, it executes all 159 predicates:                                                                                                                      
                                                                                                                                                                                                                                                           
  ```text                                                                                                                                                                                                                                                  
  for source_part in 64_source_parts:                                                                                                                                                                                                                      
      for predicate in 159_predicates:                                                                                                                                                                                                                     
          scanner(                                                                                                                                                                                                                                         
              fragments=source_part.fragments,                                                                                                                                                                                                             
              filter=predicate,                                                                                                                                                                                                                            
              use_scalar_index=True,                                                                                                                                                                                                                       
          )                                                                                                                                                                                                                                                
  ```
                                                                                                                                                                                                                                                
This produces approximately:                                                                                                                                                                                                                             
                                                                                                                                                                                                                                                         
64 source parts × 159 predicates = 10,176 Lance scanner executions                                                                                                                                                                                       
                                                                                                                                                                                                                                                         
per cache root for only the first include file.                                                                                                                                                                                                          
                                                                                                                                                                                                                                                         
In the observed run:                                                                                                       

- the first wave of 8 source parts took approximately 80–110 minutes
- each participating process remained near 100% CPU usage
- each process used approximately 38–41 GiB RSS
- each completed part contained only about 1,250 matched rows and roughly 10 KiB of row-index output
- model execution and DataLoader worker processing had not started yet

The result is therefore sparse, and the time is spent querying/loading the scalar index rather than writing the filtered-row sidecar.
The cache root was node-local, which caused the two nodes to repeat the build independently. This is an additional 2x cluster-level cost, but it does not explain the per-node behavior above: ranks within each node were already cooperating on distinct parts, and one node alone still required 80–110 minutes for its first 8 of 64 parts.

### Root cause

The current execution plan partitions work by source fragments and then repeats every bounded predicate for every source partition.

For sparse indexed inclusion, the preferable execution plan is to partition by predicate/ID group:

```
for predicate in distributed_predicate_groups:
    query the global scalar index once
    collect matching row addresses
merge, deduplicate, and order the resulting row positions
```

With 159 predicates, this would require approximately 159 global indexed queries instead of 10,176 fragment-constrained queries.

The current query count scales approximately as: `O(number_of_source_parts × number_of_predicate_batches)`

For a scalar-index-backed sparse include operation, it should ideally scale closer to: `O(number_of_predicate_batches)`

Changing the caller-side batch size only changes the constant. It does not remove the multiplication by the number of source parts. Constructing one oversized IN (...) expression would also contradict the bounded-query usage recommended in #23.

### Why this belongs to the dataset API

The caller cannot select or influence the filter-index execution plan through the current API. It already supplies bounded predicates and enables scalar-index usage. Source partitioning, scanner construction, row-address collection, cache-part
planning, deduplication, and ordering are all owned by mvp-dataset.

Therefore this is not a semantic API misuse. It is a missing performance optimization in the source-level Lance filtering implementation for the exact sparse-include use case described in this issue.

### Suggested completion criteria

The issue could be considered fully resolved when:

1. Sparse ID inclusion with a scalar/BTREE index does not execute every predicate once per source partition.
2. Query work can be distributed by predicate batch across ranks/workers.
3. Matching row addresses are merged and deduplicated while preserving the documented source-order semantics.
4. Memory usage remains bounded during merge/index construction.
5. Cache and resume behavior remain deterministic and compatible with the existing filter fingerprint.
6. A scale test or benchmark covers at least billions of source rows and tens of thousands of included IDs.

The functionality added in #23 remains valuable and correct. The request to reopen is specifically about completing the source-level performance objective stated in #21.

**Comment by flymin** [comment]
## Follow-up: recipe-local query-partitioned implementation validates the proposed execution plan

We implemented a recipe-local prototype to verify whether the bottleneck comes from Lance itself or from the filter-index execution plan.

The prototype keeps the same fundamental inputs:

- approximately 2.95 billion physical rows
- 8,323 fragments
- 81,000 unique included IDs
- 512 IDs per bounded query
- 159 query groups
- a complete scalar/BTREE index
- the same cache manager and cooperative rank participation

The main change is the dimension used to partition cache-building work.

### Query-partitioned implementation

Instead of partitioning the source into fragment ranges and executing every predicate against every range, the prototype makes each bounded ID group one independently claimable cache part.

Simplified pseudocode:

```python
ids = deduplicate(read_include_file())
include_position = {
    sample_id: position
    for position, sample_id in enumerate(ids)
}

# Sorting improves index-query locality. Final output ordering is restored later.
query_ids = sorted(ids)
query_groups = chunk(query_ids, size=512)

validate_complete_btree_index(id_field)

# CacheManager distributes these parts across participating ranks/processes.
for group in cooperatively_claimed(query_groups):
    predicate = build_in_predicate(id_field, group)

    scanner = lance_dataset.scanner(
        columns=[id_field],
        filter=predicate,
        with_row_address=True,
        scan_in_order=False,
        use_scalar_index=True,
    )

    pairs = []
    for row in scanner:
        global_row_index = row_address_to_global_index(row.row_address)
        pairs.append(
            (
                include_position[row.id],
                global_row_index,
            )
        )

    persist_query_part(pairs)

# A second cache entry merges the independently built query parts.
all_pairs = concatenate(load_all_query_parts())

# The recipe requires include-file order. The secondary physical-row key makes
# duplicate source IDs deterministic.
all_pairs.sort(key=lambda pair: (pair.include_position, pair.global_row_index))

persist_filter_index(
    global_row_index
    for _, global_row_index in all_pairs
)
```

The resulting disk-backed row index is then passed to the existing `LanceFilterIndex` and Lance iterator. Cache fingerprints include the source fingerprint, include-file fingerprint, ID field, query batch size, and ordering policy.

### Measured result

For the same 81,000-ID workload:

- query cache preparation started at `14:04:21`
- the final include index was ready at `14:14:32`
- total cold-build time was approximately 10 minutes 11 seconds
- all 81,000 requested IDs matched
- the complete index was produced successfully

By comparison, the fragment-partitioned implementation required approximately 80–110 minutes merely to finish its first 8 of 64 source partitions.

This is not just cache reuse: the prototype used a new cache kind and cache key. The improvement comes from changing the execution plan.

### Complexity difference

The merged implementation effectively performs:

```text
source_parts × query_groups
= 64 × 159
= 10,176 scanner executions
```

per cache root.

The recipe-local implementation performs approximately:

```text
query_groups
= 159 global scalar-index queries
```

per source dataset and cache root.

The experiment therefore provides direct evidence that the scalar/BTREE index is capable of serving this workload efficiently when each bounded predicate is queried globally only once.

### Why this should move into mvp-dataset

The prototype currently has to implement dataset-library responsibilities in recipe code:

- subclass `LanceDataset`
- invoke the private `_LanceSourceIterator`
- construct `LanceFilterIndex` directly
- validate BTREE coverage
- translate Lance row addresses into global row indexes
- coordinate query parts through `CacheManager`
- fingerprint and merge the resulting sidecar caches

These are source and cache implementation details rather than recipe-level concerns. The recipe should only need to provide an ID field, an include file, and possibly an ordering policy.

The prototype demonstrates a viable upstream implementation strategy:

1. Make predicate batches, rather than source fragment ranges, the cooperative cache parts for sparse indexed filters.
2. Execute each bounded predicate once against the complete scalar index.
3. Persist `(logical_order_key, global_row_index)` pairs per query part.
4. Merge and deduplicate those parts into the final disk-backed row mapping.
5. Apply the documented ordering policy when producing the final index.
6. Reuse the existing iterator, fingerprint, resume, shuffle, and selection machinery.

### Ordering note

The recipe-local prototype intentionally restores include-file order because deterministic packing must follow the supplied validation order.

The generic `filter()` API currently documents source-order preservation. An upstream implementation can retain that behavior by sorting/deduplicating the final result by `global_row_index` instead:

```python
global_row_indexes = sorted(set(
    row_index
    for query_part in query_parts
    for row_index in query_part
))
```

Therefore query-based partitioning does not require changing the public ordering semantics. Include-file order could be exposed separately by a first-class `filter_by_id(..., order="include")` option if desired.

### Remaining implementation considerations

The prototype is intentionally specialized and currently:

- requires a complete BTREE index and fails fast otherwise
- does not support Lance deletion files
- merges all matched pairs in memory
- uses a node-local cache root in the observed run

For sparse include sets, the in-memory merge is small. A general upstream implementation could use an external merge or partitioned row-index files to keep memory bounded for larger result sets.

The measured result confirms that the performance objective in #21 is achievable. The remaining work is to provide this execution plan inside `mvp-dataset` so callers do not need to reproduce internal Lance filtering and cache logic.

**Comment by GeoffreyChen777** [comment]
## mvp-engine convergence update

We have decided to remove the recipe-local Lance iterator/filter-index implementation from the KingLLM recipe and converge on the existing public `Dataset.filter()` API.

The recipe wiring now only:

- reads and deduplicates newline-delimited string IDs;
- converts include IDs into bounded `id IN (...)` predicate batches;
- converts exclude IDs into bounded `id NOT IN (...)` predicates applied through repeated `filter()` calls;
- delegates row-index construction, source-order semantics, cache publication, sharding, shuffle, and resume compatibility to `mvp-dataset`.

The recipe intentionally accepts the current `filter()` contract: matching physical rows are deduplicated and returned in Lance source order. It no longer subclasses `LanceDataset`, invokes `_LanceSourceIterator`, constructs `LanceFilterIndex`, or maintains its own query/index cache.

This means the remaining sparse-include performance problem is expected to be solved upstream without changing the mvp-engine recipe API. The desired optimization is the query-partitioned execution plan already described above: make bounded predicate batches cooperative cache parts, execute each predicate once against the global scalar/BTREE index, then merge/deduplicate global row indexes back into source order.

Keeping the public call as `dataset.filter(predicates)` while replacing the current `source_parts × predicates` build plan would let mvp-engine benefit transparently from the optimization. The KingLLM recipe path is `recipes/kingllm/kit/llm/data/`.
